### ISSUE TITLE
[no-unnecessary-condition] wrong report for variable used in iterator method.

### ISSUE THREAD

**Issue by ithater** [typescript-eslint/typescript-eslint#4501]
## Repro

[Playground](https://typescript-eslint.io/play#ts=4.5.2&sourceType=module&code=JYWwDg9gTgLgBAfTgMyhEcDkAbCATAQwGcALTAbgChKBjCAOyPjwgGV0BTAFROHoHM4AXjgAKQjAIBKYQD44Ab0pw4dRvA70YwbRyLC4AbUwBGTABosAJkwBdKiuwd4wIgDUC2YHgMwoAVw4qZUQAOg4CGhJRTW1dIksYrR0ATxkheSUVFTUmOCg9f2x4EQkCQ1jU+2psuGBkMQKiIpgZLNr8538oegdsgF8a7NcPLx8RZE8iIJD+qWCVerERz282uAB6DbgAdzQBTshYEJUCmG76OHoi7D7BkLOLuD9Aqn7gpdEXjnWt3egANb6ZB8DgnOC4GieVgwaAEfgcULTGCiTAECxwExWADMUko-SAA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuBBcSiCAhilFkmQQCZoRocgAuEHBQoyKEAF8gA&tsConfig=N4XyA)

### Code

```ts
import _ from 'lodash';

const doSomeThing = (data) => {
  const entities = ['1', '2'];
  let isValid = true;

  _.each(entities, (entity) => {
    const result = data[entity];

    if (result) {
      return;
    }

    isValid = false;
  });

  if (isValid) { // wrong report
    return null;
  }

  return true;
};

if (true) { // works fine
    localStorage.set('a', 123)
}
```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

## Expected Result
In example above, there is only 1 report about `if (true) {`.

## Actual Result
In example above, there are 2 reports about `if (true) {` and `if (isValid) {`.

## Versions

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.2` |
| `@typescript-eslint/parser`        | `5.10.2` |
| `TypeScript`                       | `4.5.2` |
| `ESLint`                           | `8.1.0` |
| `Env`                              | `web` |

**Comment by bradzacher** [comment]
There is a few issues on this but in short this isn't a false positive.

You're working your idea of it being a false positive on in depth knowledge of how `_.each` works. You know it works by immediately calling the callback synchronously.

TS does not and cannot have this info, so it assumes the callback is called in the future at some time.

So as far as TS is concerned, `isValid` is not modified synchronously.

So when we ask TS for the type of the variable, it is typed as `true`, not `boolean`.
[You can verify this by hovering the variable in your IDE](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAfTgMyhEcDkAbCATAQwGcALTAbgChKBjCAOyPjwgGV0BTAFROHoHM4AXjgAKQjAIBKYQD44Ab0pw4dRvA70YwbRyLC4AbUwBGTABosAJkwBdKiuwd4wIgDUC2YHgMwoAVw4qZUQAOg4CGhJRTW1dIksYrR0ATxkheSUVFTUmOCg9f2x4EQkCQ1jU+2psuGBkMQKiIpgZLNr8538oegdsgF8a7NcPLx8RZE8iIJD+qWCVerERz282uAB6DbgAdzQBTshYEK3sgD0AfhCVAphu+jh6Iuw+wZDb+7g-QKp+4KXRN8OOtTjtoABrfTIPgcEK4GieVgwaAEfgcULTGCiTAECxwExWADMUko-SAA).

Our information is entirely based off of the TS types. So we are as correct as they are.
